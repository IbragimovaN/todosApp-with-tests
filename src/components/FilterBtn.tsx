type Props = {
  onClick: () => void;
  disabled: boolean;
  isFilter: boolean;
};
export const FilterButton = ({ onClick, disabled, isFilter }: Props) => {
  return (
    <button
      className="button"
      onClick={onClick}
      disabled={disabled}
      data-testid="filter-button"
    >
      {isFilter ? "все задачи" : "не выполненные"}
    </button>
  );
};
