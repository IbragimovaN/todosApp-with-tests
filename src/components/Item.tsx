import { DeleteButton } from "./DeleteButton";

type Props = Task & {
  onDelete: (id: Task["id"]) => void;
  onToggle: (id: Task["id"]) => void;
};

export const Item = (props: Props) => {
  const { header } = props;

  if (header.length === 0 || header.length >= 32) {
    return null;
  }

  return (
    <li className="item-wrapper">
      <input
        type="checkbox"
        id={props.id}
        defaultChecked={props.done}
        onChange={() => props.onToggle(props.id)}
      />
      <label htmlFor={props.id} onClick={() => props.onToggle(props.id)}>
        {props.done ? <s>{props.header}</s> : props.header}
      </label>
      <DeleteButton
        disabled={!props.done}
        onClick={() => props.onDelete(props.id)}
      />
    </li>
  );
};
