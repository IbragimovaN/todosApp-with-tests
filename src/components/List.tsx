import { uncompleteCount } from "src/store/taskSlice";
import { Item } from "./Item";

type Props = {
  items: Task[];
  onDelete: (id: Task["id"]) => void;
  onToggle: (id: Task["id"]) => void;
};

export const List = ({ items, onDelete, onToggle }: Props) => {
  if (items.filter((x) => !x.done).length > 10) {
    items.pop();
  }

  return (
    <ul className="task-list tasks">
      {items.map((item) => (
        <Item {...item} key={item.id} onDelete={onDelete} onToggle={onToggle} />
      ))}
    </ul>
  );
};
