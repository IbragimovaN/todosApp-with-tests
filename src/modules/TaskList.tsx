import { useState } from "react";
import { useDispatch, useSelector } from "react-redux";
import { Empty } from "src/components/Empty";
import { FilterButton } from "src/components/FilterBtn";
import { List } from "src/components/List";
import { deleteTask, tasksSelector, toggleTask } from "src/store/taskSlice";

export const TaskList = () => {
  const items = useSelector(tasksSelector);
  const dispatch = useDispatch();
  const [isFilter, setIsFilter] = useState<boolean>(false);

  const handleDelete = (id: Task["id"]) => {
    dispatch(deleteTask(id));
  };

  const handleToggle = (id: Task["id"]) => {
    dispatch(toggleTask(id));
  };

  const handleFiltered = () => {
    setIsFilter((prev) => !prev);
  };

  return (
    <>
      <FilterButton
        onClick={handleFiltered}
        disabled={items.length === 0}
        isFilter={isFilter}
      />
      {items.length > 0 ? (
        <List
          items={isFilter ? items.filter((item) => item.done === false) : items}
          onDelete={handleDelete}
          onToggle={handleToggle}
        />
      ) : (
        <Empty />
      )}
    </>
  );
};
