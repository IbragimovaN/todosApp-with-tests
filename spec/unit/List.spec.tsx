import { render, screen } from "@testing-library/react";
import { List } from "src/components/List";
import { items } from "../utils/listForTests";
import { TaskList } from "src/modules/TaskList";
import { JestStoreProvider } from "../utils/JestStoreProvider";
import ue from "@testing-library/user-event";

const userEvent = ue.setup({
  advanceTimers: jest.advanceTimersByTime,
});

it("отображение списка задач", () => {
  const onDelete = jest.fn();
  const onToggle = jest.fn();

  const { rerender, asFragment } = render(
    <List items={items} onDelete={onDelete} onToggle={onToggle} />
  );
  const firstRender = asFragment();

  items.pop();

  rerender(<List items={items} onDelete={onDelete} onToggle={onToggle} />);
  const secondRender = asFragment();

  expect(firstRender).toMatchDiffSnapshot(secondRender);
});

it("Список содержит не больше 10 невыполненных задач", async () => {
  render(<TaskList />, {
    wrapper: JestStoreProvider,
  });

  const filterBtnEl = screen.getByTestId("filter-button");
  await userEvent.click(filterBtnEl);
  const itemsEl = screen.queryAllByRole("listitem");
  expect(itemsEl.length).toBeLessThanOrEqual(10);
});
