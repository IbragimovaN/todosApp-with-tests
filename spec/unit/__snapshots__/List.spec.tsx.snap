// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`отображение списка задач 1`] = `
"Snapshot Diff:
- First value
+ Second value

@@ -48,33 +48,7 @@
            alt="Удалить"
            src="[object Object]"
          />
        </button>
      </li>
-     <li
-       class="item-wrapper"
-     >
-       <input
-         checked=""
-         id="3"
-         type="checkbox"
-       />
-       <label
-         for="3"
-       >
-         <s>
-           выгулять собаку
-         </s>
-       </label>
-       <button
-         class="button button-with-icon"
-         data-alt="Удалить"
-         data-testid="delete"
-       >
-         <img
-           alt="Удалить"
-           src="[object Object]"
-         />
-       </button>
-     </li>
    </ul>
  </DocumentFragment>"
`;
